### PR TITLE
Add tests for common reference semantics

### PR DESCRIPTION
**Pull request by nolmoonen**
Implements test plan #472.
- No implementation supports `(un)sampled_image` or `(host_)(un)sampled_image_accessor`, but the tests are included for the sake of completeness.
- Several tests for the common reference semantics are removed as these are now placed in a central location (`common/semantics_reference.h`). The existing `common_semantics` namespace is renamed to `common_by_value_semantics` so that it is clear that this namespace only concerns the by-value semantics.
- The mutation tests of the test plan are slightly altered. For `event`, the contents of the wait list are checked instead of its size, since the implementation may keep completed events in the list. The test for `unsampled_image_accessor` is removed, as the accessor must be either `read` or `write`, not `read_write`. Hence, no modifications can be made that can also be detected.

### CODE REVIEW

**Review by ProGTX — APPROVED** [review]

File: tests/accessor/generic_accessor_semantics.cpp [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+#endif
+  sycl::buffer<int> buffer{sycl::range<1>{1}};
+
+  SECTION("mutation to copy") {
+    {
+      sycl::buffer<int> buffer_result{&result, sycl::range<1>{1}};
+      queue.submit([&](sycl::handler& cgh) {
+        auto acc_result = buffer_result.get_access(cgh);
+        sycl::accessor<int, 1> t0 = buffer.get_access(cgh);
+        cgh.single_task<kernel_name_generic<0>>([=]() {
```
Comment (keryell) [inline tests/accessor/generic_accessor_semantics.cpp:168]: useless everywhere
```suggestion
        cgh.single_task<kernel_name_generic<0>>([=] {
```

File: tests/device/device_semantics.cpp [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+};
+
+TEST_CASE("device common reference semantics", "[device]") {
+  sycl::device device = sycl_cts::util::get_cts_object::device();
+  sycl::platform platform = sycl_cts::util::get_cts_object::platform();
+  const std::vector<sycl::device> devices = platform.get_devices();
+
+  // obtain and test with a distinct device, if possible
+  if (devices.size() > 1) {
+    for (std::size_t i = 0; i < devices.size(); i++) {
```
Comment (keryell) [inline tests/device/device_semantics.cpp:55]: for-range loop?

**Review by keryell — COMMENTED** [review]
This is massive!

File: tests/accessor/generic_accessor_semantics.cpp (hunk 1 above)
Comment (nolmoonen) [reply tests/accessor/generic_accessor_semantics.cpp]: Fixed.

File: tests/device/device_semantics.cpp (hunk 2 above)
Comment (nolmoonen) [reply tests/device/device_semantics.cpp]: Fixed, in platform semantics as well.

**Comment by nolmoonen** [comment]
Force push to resolve merge conflicts.

**Comment by nolmoonen** [comment]
Force push to resolve merge conflicts (rebase).

**Review by keryell — APPROVED** [review]
Thanks.

**Comment by nolmoonen** [comment]
Force-pushed the rebase to resolve merge conflicts.

**Comment by bader** [comment]
Can't merge this one due to merge conflicts.

**Comment by nolmoonen** [comment]
Force-pushed the rebase to resolve merge conflicts. Sorry about the many force-pushes, but this is bound to happen on a PR that affects so many different files.

**Pull request merged**